lib/rbnacl/hmac/sha256.rb

```diff
@@ -22,23 +22,86 @@ class SHA256 < Auth
       sodium_constant :BYTES
       sodium_constant :KEYBYTES
 
-      sodium_function :auth_hmacsha256,
-                      :crypto_auth_hmacsha256,
-                      %i[pointer pointer ulong_long pointer]
+      sodium_function :auth_hmacsha256_init,
+                      :crypto_auth_hmacsha256_init,
+                      %i[pointer pointer size_t]
 
-      sodium_function :auth_hmacsha256_verify,
-                      :crypto_auth_hmacsha256_verify,
-                      %i[pointer pointer ulong_long pointer]
+      sodium_function :auth_hmacsha256_update,
+                      :crypto_auth_hmacsha256_update,
+                      %i[pointer pointer ulong_long]
+
+      sodium_function :auth_hmacsha256_final,
+                      :crypto_auth_hmacsha256_final,
+                      %i[pointer pointer]
+
+      # Create instance without checking key length
+      #
+      # RFC 2104 HMAC
+      # The key for HMAC can be of any length.
+      #
+      # see https://tools.ietf.org/html/rfc2104#section-3
+      def initialize(key)
+        @key = Util.check_hmac_key(key, "#{self.class} key")
+        @state = State.new
+        @authenticator = Util.zeros(tag_bytes)
+
+        self.class.auth_hmacsha256_init(@state, key, key.bytesize)
+      end
+
+      # Compute authenticator for message
+      #
+      # @params [#to_str] message message to construct an authenticator for
+      def update(message)
+        self.class.auth_hmacsha256_update(@state, message, message.bytesize)
+        self.class.auth_hmacsha256_final(@state.clone, @authenticator)
```

Comment:
This should probably be moved to `#digest` below so as to allow `#update` to be called multiple times. I'm also curious how this handles `#update` being called after it's been finalized. That should probably raise an exception.

Reply:
Hi @tarcieri. You can use this `update` multiple times. 
But the point is, `#update` method without calling this line `self.class.auth_hmacsha256_final(@state.clone, @authenticator)` every time will be rewriting the `@state`. I can add spec test to compare with `openssl` multiple usage. But I know that it's work because I test it.

Reply:
here is RbNaCl example: 
```bash
pry(main)> key = ['07c0'].pack('H*')
pry(main)> hmac = RbNaCl::HMAC::SHA256.new(key)
pry(main)> hmac.update 'abcd'
pry(main)> hmac.update '1'
pry(main)> hmac.digest
=> "G\xCC[\xBDI\x1A\xA3\x95\xCD\xA2@\xC2\xB2L\xCD\xE9\x03kV\xDA\x7F\xAB@\xBE\xB5\xA7V\x1C\x11G\xBE\xD8"
```

OpenSSL example:
```bash
pry(main)> key = ['07c0'].pack('H*')
pry(main)> hmac = OpenSSL::HMAC.new(key, d)
pry(main)> hmac.update 'abcd'
pry(main)> hmac.update '1'
pry(main)> hmac.digest
=> "G\xCC[\xBDI\x1A\xA3\x95\xCD\xA2@\xC2\xB2L\xCD\xE9\x03kV\xDA\x7F\xAB@\xBE\xB5\xA7V\x1C\x11G\xBE\xD8"
```